pkg/apis/kubeflow.org/v2alpha1/trainingruntime_types.go

```diff
@@ -15,3 +15,173 @@ limitations under the License.
+type ClusterTrainingRuntimeList struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Standard list metadata.
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// List of ClusterTrainingRuntimes.
+	Items []ClusterTrainingRuntime `json:"items"`
+}
+
+// TrainingRuntime represents a training runtime which can be referenced as part of
+// `trainingRuntimeRef` API in TrainJob. This resource is a namespaced-scoped and can be referenced
+// by TrainJob that created in the *same* namespace as the TrainingRuntime.
+type TrainingRuntime struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Standard object's metadata.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Specification of the desired TrainingRuntime.
+	Spec TrainingRuntimeSpec `json:"spec"`
+}
+
+// TrainingRuntimeList is a collection of training runtimes.
+type TrainingRuntimeList struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// Standard list metadata.
+	metav1.ListMeta `json:"metadata,omitempty"`
+
+	// List of TrainingRuntimes.
+	Items []TrainingRuntime `json:"items"`
+}
+
+// TrainingRuntimeSpec represents a specification of the desired training runtime.
+type TrainingRuntimeSpec struct {
+	// Configuration for the runtime-specific parameters, such as Torch or MPI.
+	MLSpec *MLSpec `json:"mlSpec,omitempty"`
+
+	// Number of training nodes.
+	// Defaults to 1.
+	NumNodes *int32 `json:"numNodes,omitempty"`
+
+	// JobSet configuration which will be used by TrainJob.
+	JobSetSpec *jobsetv1alpha2.JobSetSpec `json:",inline"`
+
+	// Configuration for the PodGroup to enable gang-scheduling via supported plugins.
+	PodGroupSpec *PodGroupSpec `json:"podGroupSpec,omitempty"`
+}
+
+// PodGroupSpec represents a PodGroup configuration to enable gang-scheduling.
+type PodGroupSpec struct {
+	// Plugin for the gang-scheduling.
+	Plugin GangSchedulerPlugin `json:"plugin"`
+
+	// Time threshold to schedule PodGroup for gang-scheduling.
+	ScheduleTimeoutSeconds *string `json:"scheduleTimeoutSeconds,omitempty"`
+}
+
+// GangSchedulerPlugin represents one of the supported gang-scheduling plugins.
+type GangSchedulerPlugin string
+
+const (
+	// Volcano plugin for gang-scheduling.
+	GangSchedulerPluginVolcano GangSchedulerPlugin = "volcano"
+
+	// Coscheduling plugin from the Kubernetes scheduler-plugins for gang-scheduling.
+	GangSchedulerPluginCoscheduling GangSchedulerPlugin = "coscheduling"
+)
+
+// MLSpec represents the runtime-specific configuration for various technologies.
+// One of the following specs can be set.
+type MLSpec struct {
+	// Configuration for the PyTorch runtime.
+	TorchSpec *TorchSpec `json:"torchSpec,omitempty"`
+
+	// Configuration for the MPI Runtime.
+	MPISpec *MPISpec `json:"mpiSpec,omitempty"`
+}
+
+// TorchSpec represents a PyTorch runtime configuration.
+type TorchSpec struct {
+	// Number of processes per node.
+	// This value is inserted into the `--nproc-per-node` argument of the `torchrun` CLI.
+	// Supported values: `auto`, `cpu`, `gpu`, or int value.
```

Comment:
You could probably use KubeBuilder validations for the enums here.

Reply:
As we discussed offline, we will add validations in the separate PRs.
Tracking issue:
- https://github.com/kubeflow/training-operator/issues/2209
- https://github.com/kubeflow/training-operator/issues/2219